server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -0,0 +1,372 @@
+        // todo parallel upload
+        // any validations before/after upload ?
+        for (IndexMetadata indexMetadata : clusterState.metadata().indices().values()) {
+            // 123456789012_test-cluster/cluster-state/dsgYj10Nkso7/index/ftqsCnn9TgOX/metadata_4_1690947200
+            final String indexMetadataKey = writeIndexMetadata(
+                clusterState.getClusterName().value(),
+                clusterState.getMetadata().clusterUUID(),
+                indexMetadata,
+                indexMetadataFileName(indexMetadata)
+            );
+            final UploadedIndexMetadata uploadedIndexMetadata = new UploadedIndexMetadata(
+                indexMetadata.getIndex().getName(),
+                indexMetadata.getIndexUUID(),
+                indexMetadataKey
+            );
+            allUploadedIndexMetadata.add(uploadedIndexMetadata);
+        }
+        final ClusterMetadataManifest manifest = uploadManifest(clusterState, allUploadedIndexMetadata, false);
+        final long durationMillis = relativeTimeMillisSupplier.getAsLong() - startTimeMillis;
+        if (durationMillis >= slowWriteLoggingThreshold.getMillis()) {
+            logger.warn(
+                "writing cluster state took [{}ms] which is above the warn threshold of [{}]; " + "wrote full state with [{}] indices",
+                durationMillis,
+                slowWriteLoggingThreshold,
+                allUploadedIndexMetadata.size()
+            );
+        } else {
+            // todo change to debug
+            logger.info(
+                "writing cluster state took [{}ms]; " + "wrote full state with [{}] indices",
+                durationMillis,
+                allUploadedIndexMetadata.size()
+            );
+        }
+        return manifest;
+    }
+
+    /**
+     * This method uploads the diff between the previous cluster state and the current cluster state. The previous manifest file is needed to create the new
+     * manifest. The new manifest file is created by using the unchanged metadata from the previous manifest and the new metadata changes from the current cluster
+     * state.
+     *
+     * @return The uploaded ClusterMetadataManifest file
+     */
+    @Nullable
+    public ClusterMetadataManifest writeIncrementalMetadata(
+        ClusterState previousClusterState,
+        ClusterState clusterState,
+        ClusterMetadataManifest previousManifest
+    ) throws IOException {
+        final long startTimeMillis = relativeTimeMillisSupplier.getAsLong();
+        if (clusterState.nodes().isLocalNodeElectedClusterManager() == false) {
+            logger.error("Local node is not elected cluster manager. Exiting");
+            return null;
+        }
+        assert previousClusterState.metadata().coordinationMetadata().term() == clusterState.metadata().coordinationMetadata().term();
+        assert REMOTE_CLUSTER_STATE_ENABLED_SETTING.get(settings) == true : "Remote cluster state is not enabled";
```

Comment:
we dont need assert here as you moved it to constructor

Reply:
Removed